src/torrent.cpp

```diff
@@ -6141,8 +6144,10 @@ bool is_downloading_state(int const st)
 
 		auto const& f = m_torrent_file->file_merkle_tree(req.file);
 
-		int const base_layer_idx = merkle_num_layers(merkle_num_leafs(
-			aux::numeric_cast<int>(f.size()))) - req.base;
```

Comment:
this was not correct (afaict). `merkle_num_leafs()` is not very well named. it takes the number *pieces* and returns that rounded up to an even power of two. So, its input is the number of pieces or blocks (blocks in this case), not the total size of the tree.

Reply:
in fact, I'm tempted to turn the free `merkle_*` functions into a merkle tree class, that owns the tree representation as well. That would fix usability issues like this and hide the internal `aux::vector`. I think such type would have to be public though, as it's part of `add_torrent_params` and `torrent_status` (I think). What do you think about that?

Reply:
At first I was skeptical, but after looking at how and where these functions are used I think it could be a significant improvement. In hindsight functions like `file_storage::file_first_piece_node` don't belong in `file_storage`, they should be part of the merkle tree class. It could also improve type safety by tying the tree traversal functions to a custom iterator rather than using bare ints.